text/0517-io-os-reform.md

```diff
@@ -447,7 +457,173 @@ counts, arguments to `main`, and so on).
+  (likely out of tree).
+
+* **The error type**. The traits currently use `IoResult` in their
+  return types, which ties them to `IoError` in particular. Besides
+  being an unnecessary restriction, this type prevents `Reader` and
+  `Writer` (and various adapters built on top of them) from moving to
+  `libcore` -- `IoError` currently requires the `String` type.
+
+  With associated types, there is essentially no downside in making
+  the error type generic.
+
+With those general points out of the way, let's look at the details.
+
+### Nonatomic results
+[Nonatomic results]: #nonatomic-results
+
+To clarity dealing with nonatomic operations and improve their
+ergonomics, we introduce some new types into `std::error`:
+
+```rust
+// The progress so far (T) paired with an err (Err)
+struct PartialResult<T, Err>(T, Err);
+
+// An operation that may fail after having made some progress:
+// - S is what's produced on complete success,
+// - T is what's produced if an operation fails part of the way through
+type NonatomicResult<S, T, Err> = Result<S, PartialResult<T, Err>>;
+
+// Ergonomically throw out the partial result
+impl<T, Err> FromError<PartialResult<T, Err> for Err { ... }
+```
+
+The `NonatomicResult` type (which could use a shorter name)
+encapsulates the common pattern of operations that may fail after
+having made some progress. The `PartialResult` type then returns the
+progress that was made along with the error, but with a `FromError`
+implementation that makes it trivial to throw out the partial result
+if desired.
+
+### `Reader`
+[Reader]: #reader
+
+The updated `Reader` trait (and its extension) is as follows:
+
+```rust
+trait Reader {
+    type Err; // new associated error type
+
+    // unchanged except for error type
+    fn read(&mut self, buf: &mut [u8]) -> Result<uint, Err>;
+
+    // these all return partial results on error
+    fn read_to_end(&mut self) -> NonatomicResult<Vec<u8>, Vec<u8>, Err> { ... }
+    fn read_to_string(&self) -> NonatomicResult<String, Vec<u8>, Err> { ... }
+    fn read_at_least(&mut self, min: uint,  buf: &mut [u8]) -> NonatomicResult<uint, uint, Err>  { ... }
+}
+
+// extension trait needed for object safety
+trait ReaderExt: Reader {
+    fn bytes(&mut self) -> Bytes<Self> { ... }
+    fn chars<'r>(&'r mut self) -> Chars<'r, Self, Err> { ... }
+
+    ... // more to come later in the RFC
+}
+impl<R: Reader> ReaderExt for R {}
+```
+
+#### Removed methods
+
+The proposed `Reader` trait is much slimmer than today's. The vast
+majority of removed methods are parsing/deserialization, which were
+discussed above.
+
+The remaining methods (`read_exact`, `push`, `push_at_least`) were
+removed largely because they are *not memory safe*: they involve
+extending a vector's capacity, and then *passing in the resulting
+uninitialized memory* to the `read` method, which is not marked
+`unsafe`! Thus the current design can lead to undefined behavior in
+safe code.
```

Comment:
I don't see how. I doubt that memory returned by an allocator will be treated as undef by LLVM. And even if this is the case, it should be easily possible to tell LLVM that the memory is not undef. Apart from this issue, this is equivalent to passing the same slice multiple times to `read` without zeroing in between.

Again: Please explain how this can lead to UB.